(PUP-9065) Support for write/read of deferred functions

tvpartytonight opened the pull request.

tvpartytonight commented on lib/puppet/pops/evaluator/deferred_resolver.rb:
```diff
@@ -14,13 +14,12 @@ class DeferredResolver
-  def self.resolve_and_replace(node, facts, catalog)
-    compiler = Puppet::Parser::ScriptCompiler.new(node.environment, node.name)
+  def self.resolve_and_replace(facts, catalog)
+    compiler = Puppet::Parser::ScriptCompiler.new(catalog.environment_instance, catalog.name)
```
In my testing, node.name and catalog.name where interchangeable; is there a case where they might not be?

hlindberg replied:
If the node changed name since it got the catalog?
The change is ok if it is known that catalog.environment_instance is the same thing and catalog.name is the same as node.name

MikaelSmith commented on lib/puppet/configurer.rb:
```diff
@@ -109,8 +109,17 @@ def convert_catalog(result, duration, options = {})
+        # Perhaps the prepare_and_retrieve_catalog_from_cache method
+        # could check a Puppet setting to determine if it should run facter or
+        # not?
+        Puppet::Pops::Evaluator::DeferredResolver.resolve_and_replace({}, result)
```
I'm not clear why facts may not have been collected with a cached catalog run. They're still used by providers.

tvpartytonight replied:
If I run `puppet agent -t --use_cached_catalog`, the `convert_catalog` is called before `get_facts` is. Given Henrik's comments, we might just want to ensure that facts are always run earlier so that the user cannot apply a catalog without facts.

hlindberg commented on lib/puppet/configurer.rb:
```diff
@@ -109,8 +109,17 @@ def convert_catalog(result, duration, options = {})
+        Puppet::Pops::Evaluator::DeferredResolver.resolve_and_replace(facts, result)
+      else
+        # During a cached_catalog run, facts may not have been collected; even in
+        # that case, we should still attempt to reolve_and_replace deferred
```
typo `reolve`

hlindberg commented on lib/puppet/configurer.rb:
```diff
@@ -109,8 +109,17 @@ def convert_catalog(result, duration, options = {})
+      else
+        # During a cached_catalog run, facts may not have been collected; even in
+        # that case, we should still attempt to reolve_and_replace deferred
+        # functions and send an empty hash for facts parameter.
```
The result would be surprising - if someone actually needs a fact value they would get nil.
So - question why does this happen and should something other than resolving deferred values be done?
This seems like a bad idea that will lead to hard to figure out problems for users.

Depending on why this happens it should either get the current facts, or it should error.

tvpartytonight replied:
I think the solution here is to call `get_facts` in `prepare_and_retrieve_catalog_from_cache`, and that way convert_catalog method won't change much, it will just not check for a node definition, and it won't try to resolve a catalog without facts.

hlindberg commented on lib/puppet/pops/serialization/to_data_converter.rb:
```diff
@@ -108,6 +108,10 @@ def to_data(value)
+      elsif value.instance_of?(Puppet::Pops::Types::TypeFactory.deferred.implementation_class)
+        process(value) do
+          { PCORE_TYPE_KEY => PCORE_TYPE_DEFERRED, 'name' => value.name, 'arguments' => value.arguments }
+        end
```
I don't think this is needed now since Deferred is an Object. When I tried just calling the to_data_converter I get the expected hash with `{__ptype: "Deferred", ... }`

hlindberg replied:
ha, but not when @rich_data is false

tvpartytonight replied:
Without this code, I get the warning that it is stringifying the deferred class when the `unknown_to_data(value)` method runs during an agent run, specifically when it is caching the catalog.

hlindberg replied:
Are you sure `rich_data` is on when caching the catalog? If not it will destroy a catalog with rich data.
The method does this:
```
def unknown_to_data(value)
  @rich_data ? value_to_data_hash(value) : unknown_to_string_with_warning(value)
end
```
So, it should just convert the object to its data hash.

I am not sure why the special handling of Sensitive is there - it seems to always produce a hash with __ptype => Sensitive even if not using rich_data.

hlindberg commented on lib/puppet/pops/serialization.rb:
```diff
@@ -28,6 +28,9 @@ class SerializationError < Puppet::Error
   PCORE_LOCAL_REF_SYMBOL = 'LocalRef'.freeze
+
+  # Type key used for deferred functions
+  PCORE_TYPE_DEFERRED = 'Deferred'.freeze
```
Don't think this is needed

tvpartytonight commented:
@hlindberg I think I have addressed all the feedback on this PR; do you see any more things to address?

hlindberg commented:
I would like to see a test with a catalog containing rich data (not just Deferred values) that gets cached and then verified to be the same after read back from cache.
For example:
```puppet
notify { test: message => /.*/ }
```
Should have a catalog with the resource like this in Json:
```json
    {
      "type": "Notify",
      "title": "test",
      "tags": [
        "notify",
        "test"
      ],
      "line": 1,
      "exported": false,
      "parameters": {
        "message": {
          "__ptype": "Regexp",
          "__pvalue": ".*"
        }
      }
    }
```

tvpartytonight commented:
This is blocked on #7054 ; should reevaluate if this is still needed after that is merged.

hlindberg commented:
Thought about the issues here. I think the cached catalog should always be written using rich_data turned on - that way the written catalog will always be the same as the original (this since writing an initially non rich catalog with rich_data flag on will produce a true copy of the catalog, and it will be correct for a rich data catalog as well).

puppetcla commented:
CLA signed by all contributors.

tvpartytonight commented:
I've removed the pcore changes that erroneously converted deferred types into hashes; in my testing now, the puppet agent can read in cached catalogs that have the deferred type in them: 

```
/usr/src/puppet # puppet agent -t --rich_data
Info: Using configured environment 'production'
Info: Retrieving pluginfacts
Info: Retrieving plugin
Info: Retrieving locales
Info: Caching catalog for agent.docker
Info: Applying configuration version '1536604553'
Notice: {:foo=>"bar"}
Notice: /Stage[main]/Main/Node[default]/Notify[example]/message: changed [redacted] to [redacted]
Notice: Applied catalog in 0.02 seconds

/usr/src/puppet # puppet agent -t --rich_data --use_cached_catalog
Info: Using cached catalog from environment 'production'
Info: Applying configuration version '1536604553'
Notice: {:foo=>"bar"}
Notice: /Stage[main]/Main/Node[default]/Notify[example]/message: changed [redacted] to [redacted]
Notice: Applied catalog in 0.02 seconds

```

I'd still like to add a test to this, but my comment above that this was blocked by #7054 is not true...I think.

tvpartytonight commented:
jenkins please test this

tvpartytonight commented:
jenkins please test this

hlindberg commented on acceptance/tests/pluginsync/feature/pluginsync_should_sync_features.rb:
```diff
@@ -111,7 +111,8 @@ def bogus_function()
 
         step 'run the agent and verify that it loaded the feature' do
-          on(agent, puppet("agent -t --libdir='#{agent_lib_dir}' --server #{master} --environment '#{tmp_environment}'"),
+          require 'pry-byebug';binding.pry
```
debugger stuff...

tvpartytonight commented:
jenkins please test this
